man/man1/snapfits.1
.TH SNAPFITS 1NEMO "19 March 1990"
.SH NAME
snapfits \- convert a snapshot file to a fits file.
.SH SYNOPSIS
.PP
\fBsnapfits in=\fPsnapshot \fBout=\fPfitsfile [parameter=value]
.SH DESCRIPTION
\fBsnapfits\fP converts a snapshot to a fits file, in
the random groups format (See: \fIxxx\fP). See 
the section on 'FITS KEYWORDS' below for a more
detailed description. The kind of
parameters written can be choosen
with the \fBoptions\fP keyword (see below). The
default is 7 (masses, positions and velocities for a 3-dimensional
simulation). 
.PP
Note: Only one snapshot can be written. The time of the
snapshot is not saved yet.
.SH PARAMETERS
The following parameters are the ones whose meaning is different
from the one in \fBsnapplot\fP:
.TP 20
\fBin=\fIin-file\fP
input file, must be in \fIsnapshot(5NEMO)\fP format [no default].
.TP
\fBout=\fIout-file\fP
output file, will be in \fIfits(5NEMO)\fP format [no default].
.TP
\fBoptions=\fIopt1[,opt2...]\fP
List of options, separated by comma's (no spaces),  of the
items to be saved in the output FITS file.
They can be any combination of: \fImass, phase (pos, vel), phi and acc\fP.
Note that the order of items written is fixed, as described above.
Even for \fIoptions=pos,mass\fP the mass will be preceded by the 2/3 
positions of the particle. [default: \fBmass,phase\fP].
.TP
\fBcomment=\fIcomment\fP
One comment line with up to about [70] characters can be used to identify
this FITS file. [default: empty].
.SH FITS KEYWORDS
The data is written using the random groups format in FITS without
any primary data matrix.
Hence NAXIS1=0, PGROUP the number of items per particle to be output
(7 being the default), and GCOUNT the total number of bodies in a snapshot.
The external format of the data is IEEE floating point (BITPIX = -32 or
-64, depending on precision required). There will be no scaling
(default PZEROnnn and PSCALnnn), whereas the axes are labeled using
the PTYPEnnn keyword. They can be any of MASS, POSn, VELn, PHI and
ACCn (n=1,..NDIM).
.SH BUGS
There is really no use for this program, since the general way
to transport data is using the general ascii format (see \fIatos(1NEMO)\fP or
\fIsnapprint(1NEMO)\fP) or ascii structured files (see \fItsf(1NEMO)\fP).
For large amounts of datafiles this fits format may
actually become useful. It was merely created for the
authors amusement and to show
how the FITS format could be used to export N-body data.
.PP
There is no program \fIfitssnap\fP yet. As the need arises, and
this fits-i/o format becomes popular, the author may be poked 
to create one.
.SH "SEE ALSO"
atos(1NEMO), stoa(1NEMO), tsf(1NEMO), rsf(1NEMO), snapprint(1NEMO)
.SH AUTHOR
Peter Teuben
.SH FILES
.nf
.ta +2.5i
src/pjt/image	snapfits.c, snapfits.1
.fi
.SH "UPDATE HISTORY"
.nf
.ta +1.0i +4.0i
19-mar-90	V1.0: Created	PJT
.fi


src/nbody/fits/snapfits.c
/* 
 *  SNAPFITS:  write a snapshot into fits format, 
 *              in random groups format that is
 *             
 *      19-mar-90  V1.0 -- created as a toy model     PJT
 *			Must be changed to 3DTABLE to be standard FITS
 */

#include <stdinc.h>             /* also gets <stdio.h>  */
#include <getparam.h>
#include <vectmath.h>           /* otherwise NDIM undefined */
#include <history.h>
#include <filestruct.h>

#include <snapshot/body.h>      /* snapshot's */
#include <snapshot/snapshot.h>
#include <snapshot/get_snap.c>

#include <fits.h>

char *defv[] = {
        "in=???\n               Input filename (a snapshot)",
        "out=???\n              Output filename (a fits file)",
        "options=mass,phase\n   Options for output",
        "comment=\n             Additional comments for FITS file",
        "VERSION=1.0\n          30-mar-90 PJT",            
        NULL,
};

static stream  instr, outstr;                          /* file streams */

                /* SNAPSHOT INTERFACE */
static int    nbody, bits;
static real   tnow;
static Body   *btab = NULL;
                /* FITS interface */
struct fits_header fh;
static char *comments[] = {NULL, NULL};        /* place to save ONE comment */

#define MAXNEEDS 10
static int needs[MAXNEEDS];     /* 0=mass, 1=pos, 2=vel 3=phi, 4=acc */


main (argc, argv)
int argc;
char *argv[];
{
    initparam (argv, defv);                     /* get user thingos */

    instr = stropen (getparam ("in"), "r");     /* open snapshot */
    get_history(instr);
    get_snap(instr,&btab,&nbody,&tnow,&bits);
    strclose(instr);

    outstr = stropen(getparam("out"), "w");
    w_header(outstr,&fh,nbody,getparam("options"),getparam("comment"));
    w_data(outstr,&fh,nbody,btab);
}

need_pars(options)
string options;
{
    int i, ntot;
    bool scanopt();
    for (i=0; i<MAXNEEDS; i++)
	needs[i] = 0;
    needs[0] = (scanopt(options,"mass") ? 1 : 0);
    if (scanopt(options,"phase"))
        needs[1] = needs[2] = NDIM;
    else
        needs[1] = needs[2] = 0;
    if (scanopt(options,"pos"))         /* catch pos,vel <-> phase */
        needs[1] = NDIM;
    if (scanopt(options,"vel"))
        needs[2] = NDIM;
    needs[3] = (scanopt(options,"phi") ? 1 : 0);
    needs[4] = (scanopt(options,"acc") ? NDIM : 0);

    for (ntot=0, i=0; i<MAXNEEDS; i++)
        ntot += needs[i];

    if (ntot==0)
        error("options musyt contain at least one of: mass,phase,pos,vel\n");
    else
        dprintf(0,"Options=%s outputs %d items per body\n",options,ntot);

    return(ntot);
}

w_header(ostr,fh,nbody,options,comment)
stream ostr;
struct fits_header *fh;
char *options, *comment;
int nbody;
{   
    static int naxis[1] = { 0 };    /* random groups ! */
    int i, n, ni;
    string *ptype, copxstr();
    char pname[9];

    fts_zero(fh);          /* clear primary header */

    fh->simple = 1;          /* set various primary header stuff */
    fh->groups = 1;
    fh->bitpix = -64;        /* IEEE double precision FOR NOW */
    fh->naxis = 1;
    fh->naxisn = naxis;      /* point to the array */
    fh->pcount = n = need_pars(options);
    fh->gcount = nbody;
    fh->history = ask_history();    /* fill in some history */
    if (*comment != NULL) {         /* were there any comments ? */
        comments[0] = comment;        
        fh->comment = comments;
    }
    
    ptype = (char **) malloc(n*sizeof(char *));
    ni = 0;
    if (needs[0]>0)
        ptype[ni++] = copxstr("MASS",sizeof(char));
    if (needs[1]>0)
        for (i=0; i<NDIM; i++) {
            sprintf(pname,"POS%1d",i+1);
            ptype[ni++] = copxstr(pname,sizeof(char));
        }
    if (needs[2]>0)
        for (i=0; i<NDIM; i++) {
            sprintf(pname,"VEL%1d",i+1);
            ptype[ni++] = copxstr(pname,sizeof(char));
        }
    if (needs[3]>0)
        ptype[ni++] = copxstr("PHI",sizeof(char));
    if (needs[4]>0)
        for (i=0; i<NDIM; i++) {
            sprintf(pname,"ACC%1d",i+1);
            ptype[ni++] = copxstr(pname,sizeof(char));
        }
    for (ni=0; ni<n; ni++)
        dprintf(0,"PTYPE%d = %s\n",ni+1,ptype[ni]);
    fh->ptypen = ptype;
    
    
    fts_whead(ostr,fh);      /* write the header */
}

w_data(ostr,fh,nbody,btab)
stream ostr;
struct fits_header *fh;
int nbody;
Body *btab;
{
    int i, nw, need_mass,need_pos,need_vel, ntot, isize;
    Body *bp;
    double out[32];   /* scratch array for buffering up a body's parameters */
    char null = NULL;
    
    ntot = 0;
    isize = sizeof(double);
    for (bp=btab;bp<btab+nbody;bp++) {
        nw = 0;
        if (needs[0]>0)
            out[nw++] = Mass(bp);
        if (needs[1]>0)
            for (i=0; i<NDIM; i++)
                out[nw++] = Pos(bp)[i];
        if (needs[2]>0)
            for (i=0; i<NDIM; i++)
                out[nw++] = Vel(bp)[i];
        if (needs[3]>0)
            out[nw++] = Phi(bp);
        if (needs[4]>0)
            for (i=0; i<NDIM; i++)
                out[nw++] = Acc(bp)[i];
            
        fts_wdata(ostr,fh,nw*isize,out);
        ntot += nw*isize;
    }
    /* make sure tail end filled with zero's */
    i = FTSBLKSIZ - (ntot % FTSBLKSIZ);
    if (i==FTSBLKSIZ) i=0;
    dprintf(0,"Wrote %d bytes, flushing %d zero's\n",ntot,i);
    while (i-- > 0)
       fts_wdata(ostr,fh,1,&null);
    
}
